From 2d155eb4ecdc4cc134bc3518c3405573df48c247 Mon Sep 17 00:00:00 2001
From: Alessandro Decina <alessandro.decina@collabora.co.uk>
Date: Thu, 8 Sep 2011 13:13:39 +0200
Subject: [PATCH 19/19] v4l2sink: more cropping fixes

Rework the previous commit (aa2dd29). Handle all the previous cases
and fix the case in which the pipeline contains elements that answers
the BUFFERS query but need no padding.
---
 sys/v4l2/gstv4l2sink.c |   35 ++++++++++++++++-------------------
 1 files changed, 16 insertions(+), 19 deletions(-)

diff --git a/sys/v4l2/gstv4l2sink.c b/sys/v4l2/gstv4l2sink.c
index 934c591..728008e 100644
--- a/sys/v4l2/gstv4l2sink.c
+++ b/sys/v4l2/gstv4l2sink.c
@@ -521,6 +521,9 @@ gst_v4l2sink_sync_overlay_fields (GstV4l2Sink * v4l2sink)
 static void
 gst_v4l2sink_sync_crop_fields (GstV4l2Sink * v4l2sink)
 {
+  if (!v4l2sink->crop_fields_set)
+    return;
+
   if (GST_V4L2_IS_OPEN (v4l2sink->v4l2object)) {
 
     gint fd = v4l2sink->v4l2object->video_fd;
@@ -540,10 +543,6 @@ gst_v4l2sink_sync_crop_fields (GstV4l2Sink * v4l2sink)
         v4l2sink->crop.top, v4l2sink->crop.left,
         v4l2sink->crop.width, v4l2sink->crop.height);
 
-    if (v4l2sink->crop_fields_set == 0) {
-      crop.c.width = GST_VIDEO_SINK_WIDTH (v4l2sink);
-      crop.c.height = GST_VIDEO_SINK_HEIGHT (v4l2sink);
-    }
     if (v4l2sink->crop_fields_set & RECT_TOP_SET)
       crop.c.top = v4l2sink->crop.top;
     if (v4l2sink->crop_fields_set & RECT_LEFT_SET)
@@ -872,6 +871,17 @@ gst_v4l2sink_set_caps (GstBaseSink * bsink, GstCaps * caps)
     gst_caps_unref (caps);
     return FALSE;
   }
+    
+  /* clear top/left crop values.. otherwise by default display will try
+   * to center, rather than scale, the image if it is too big to fit on
+   * display
+   */
+  v4l2sink->crop.top = v4l2sink->crop.left = 0;
+  v4l2sink->crop_fields_set |= RECT_TOP_SET | RECT_LEFT_SET;
+  v4l2sink->crop.width = w;
+  v4l2sink->crop_fields_set |= RECT_WIDTH_SET;
+  v4l2sink->crop.height = h;
+  v4l2sink->crop_fields_set |= RECT_HEIGHT_SET;
 
   /* query to find if anyone upstream using these buffers has any
    * minimum requirements:
@@ -899,23 +909,10 @@ gst_v4l2sink_set_caps (GstBaseSink * bsink, GstCaps * caps)
     }
 
     gst_query_parse_buffers_dimensions (query, &min_width, &min_height);
-    if (min_width > w) {
-      v4l2sink->crop.width = w;
-      v4l2sink->crop_fields_set |= RECT_WIDTH_SET;
+    if (min_width > w)
       w = min_width;
-    }
-    if (min_height > h) {
-      v4l2sink->crop.height = h;
-      v4l2sink->crop_fields_set |= RECT_HEIGHT_SET;
+    if (min_height > h) 
       h = min_height;
-    }
-
-    /* clear top/left crop values.. otherwise by default display will try
-     * to center, rather than scale, the image if it is too big to fit on
-     * display
-     */
-    v4l2sink->crop.top = v4l2sink->crop.left = 0;
-    v4l2sink->crop_fields_set |= RECT_TOP_SET | RECT_LEFT_SET;
   }
   gst_query_unref (query);
 
-- 
1.7.1

